File: src/main/org/nlogo/api/Options.java
package org.nlogo.api;

import java.util.ArrayList;
import java.util.List;

public strictfp class Options<T> {

  private final List<Option> options = new ArrayList<Option>();
  private Option current = null;

  public String chosenName() {
    return current.name;
  }

  public T chosenValue() {
    return current.value;
  }

  ///

  public void addOption(String name, T value) {
    options.add(new Option(name, value));
  }

  ///

  public List<String> getNames() {
    List<String> result = new ArrayList<String>();
    for (Option option : options) {
      result.add(option.name);
    }
    return result;
  }

  public List<T> getValues() {
    List<T> result = new ArrayList<T>();
    for (Option option : options) {
      result.add(option.value);
    }
    return result;
  }

  ///

  public void selectByName(String s) {
    for (Option option : options) {
      if (option.name != null && option.name.equals(s)) {
        current = option;
        return;
      }
    }
    throw new IllegalArgumentException("not found: " + s);
  }

  public void selectValue(T obj) {
    for (Option option : options) {
      if (option.value != null && option.value.equals(obj)) {
        current = option;
        return;
      }
    }
    throw new IllegalArgumentException("not found: " + obj);
  }

  private class Option {
    final String name;
    final T value;

    Option(String name, T value) {
      this.name = name;
      this.value = value;
    }
  }

}

File: src/main/org/nlogo/api/Options.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.api

import scala.{ Option => _ }
import collection.mutable.ArrayBuffer

class Options[T] {

  private case class Option(name: String, value: T)

  private val choices = ArrayBuffer[Option]()
  private var current: scala.Option[Option] = None

  def chosenName = current.get.name
  def chosenValue = current.get.value

  def addOption(name: String, value: T) {
    choices += Option(name, value)
  }

  def names: List[String] =
    choices.map(_.name).toList

  def values: List[T] =
    choices.map(_.value).toList

  def selectByName(s: String) {
    current = choices.find(_.name == s)
  }

  def selectValue(obj: T) {
    current = choices.find(_.value == obj)
  }

}
